src/bvh/ordered_traverse.rs

```diff
@@ -0,0 +1,269 @@
+use nalgebra::{Scalar, RealField};
+use num::traits::float::Float;
+
+use crate::aabb::{Bounded};
+use crate::bvh::{BvhNode, Bvh};
+use crate::ray::Ray;
+
+
+#[derive(Debug, Clone, Copy)]
+enum RestChild { 
+    Left, 
+    Right, 
+    None 
+}
+
+
+/// Iterator to traverse a [`Bvh`] in order from nearest [`Aabb`] to farthest for ray, 
+/// without memory allocations
+#[allow(clippy::upper_case_acronyms)]
+pub struct BvhOrderedTraverseIterator<'bvh, 'shape, T: Scalar + Copy, const D: usize, Shape: Bounded<T, D>> {
+    /// Reference to the Bvh to traverse
+    bvh: &'bvh Bvh<T, D>,
+    /// Reference to the input ray
+    ray: &'bvh Ray<T, D>,
+    /// Reference to the input shapes array
+    shapes: &'shape [Shape],
+    /// Traversal stack. 4 billion items seems enough?
+    stack: [(usize, RestChild); 32],
+    /// Position of the iterator in bvh.nodes
+    node_index: usize,
+    /// Size of the traversal stack
+    stack_size: usize,
+    /// Whether or not we have a valid node (or leaf)
+    has_node: bool,
+}
+
+
+impl<'bvh, 'shape,  T, const D: usize,  Shape: Bounded<T, D>> 
+    BvhOrderedTraverseIterator<'bvh, 'shape, T, D, Shape> 
+where
+    T: Scalar + Copy + Scalar + RealField + Float
```

Comment:
```suggestion
    T: Scalar + Copy + Scalar + RealField
```

I don't think float is needed (?)

Reply:
In this impl Ray<T, D> required T: Float for `intersection_slice_for_aabb()` method